### DISCUSSION TITLE
How to add GAP as a custom language when generating math output?

### DISCUSSION THREAD

**Discussion by kiryph** [leo-colisson/robust-externalize#7] [Q&A] [answered]
> ### What is [GAP](https://www.gap-system.org/)?
> GAP is a system for computational discrete algebra, with particular emphasis on [Computational Group Theory](https://www.gap-system.org/Doc/references.html). GAP provides a [programming language](https://docs.gap-system.org/doc/ref/chap4.html#X7FE7C0C17E1ED118), a library of thousands of functions implementing algebraic algorithms written in the GAP language as well as large [data libraries](https://www.gap-system.org/Datalib/datalib.html) of algebraic objects.

With the help of the [GAP package typeset](https://github.com/gap-packages/typeset), you can, for example, conveniently generate lists of matrix generators of standard groups as LaTeX code:

```
gap> Typeset(SpaceGroupIT(2,17));
\left\langle \left(\begin{array}{rrr}
0 & 1 & 0 \\
-1 & -1 & 0 \\
0 & 0 & 1 \\
\end{array}\right),\left(\begin{array}{rrr}
-1 & 0 & 0 \\
0 & -1 & 0 \\
0 & 0 & 1 \\
\end{array}\right),\left(\begin{array}{rrr}
0 & -1 & 0 \\
-1 & 0 & 0 \\
0 & 0 & 1 \\
\end{array}\right),\left(\begin{array}{rrr}
1 & 0 & 0 \\
0 & 1 & 0 \\
1 & 0 & 1 \\
\end{array}\right),\left(\begin{array}{rrr}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 1 & 1 \\
\end{array}\right) \right\rangle
```

So for this question I am not interested in generating images but LaTeX code which can be input into a LaTeX document.

I have come so far that it works, but the content in a math environment is repeated twice

```tex
\documentclass{article}
\usepackage{robust-externalize}

\begin{PlaceholderFromCode}{__MY_GAP_TEMPLATE__}
LoadPackage("typeset");;
__ROBEXT_MAIN_CONTENT__
\end{PlaceholderFromCode} 

\robExtConfigure{
  new preset={gap}{
    set compilation command={/usr/local/bin/gap -A -b --nointeract "__ROBEXT_SOURCE_FILE__" > "__ROBEXT_OUTPUT_PREFIX__-out.tex" && touch "__ROBEXT_OUTPUT_PDF__"},
    % DOC: 
    % Even if you do not plan to output a pdf file, you should still create
    % __ROBEXT_OUTPUT_PDF__  at the end of the compilation so that this library
    % can know whether the compilation succeeded. 
    set template={__MY_GAP_TEMPLATE__},
    custom include command={%
      \evalPlaceholder{%
        \input{__ROBEXT_CACHE_FOLDER____ROBEXT_OUTPUT_PREFIX__-out.tex}%
      }%
    },
    print verbatim if no externalization,
  }
} 

\begin{document}

\begin{equation}
  \begin{CacheMeCode}{gap}
x := [ [ 0, 0, 0, 1 ], [ 0, 0, 1, 0 ], [ 0, 1, 0, 0 ], [ 1, 0, 0, 0 ] ];;
Typeset(x);
  \end{CacheMeCode}
\end{equation}

% ISSUE content is inserted two times.


% IMPORTANT NOTE FROM DOCUMENTATION:
% Due to fundamental LATEX restrictions, it is impossible to use CacheMeCode
% inside a macro or some environments as LATEX will strip all lines containing a
% percent character for instance.
\begin{PlaceholderFromCode}{__TMP__} 
x := [ [ 0, 0, 0, 2 ], [ 0, 0, 2, 0 ], [ 0, 2, 0, 0 ], [ 2, 0, 0, 0 ] ];;
Typeset(x);
\end{PlaceholderFromCode}
\begin{equation}
  M = \cacheMe[gap]{__TMP__}
\end{equation}
% ISSUE content is still inserted two times.

\end{document}
```

The result looks as following:

![Screenshot 2023-10-29 at 11 48 52](https://github.com/leo-colisson/robust-externalize/assets/371756/08fdf5d0-a84b-4335-abae-178512e77d44)

The generated files for inclusion do not contain the result twice:
```
❯ cat robustExternalize/robExt-4D33E398C1F125AA0C26395BA9C52CA3-out.tex
\left(\begin{array}{rrrr}
0 & 0 & 0 & 1 \\
0 & 0 & 1 & 0 \\
0 & 1 & 0 & 0 \\
1 & 0 & 0 & 0 \\
\end{array}\right)
```

- **What did I do incorrectly so that the compilation result is inserted twice?**

- ~~And I also would like to run GAP in parallel similar to the FAQ entry in the documentation. However, the file `robust-ext-gap-robExt-compile-missing-figures.sh` does not contain my custom compilation command. Do I need to add something to the definition of the preset so that the shell file is filled with the GAP commands?~~ Misunderstanding on my side. Maybe the filename could be more general: `...-robExt-compile-missing-externs`.

Minor improvements would be as well:

- Customize file extension of  `__ROBEXT_SOURCE_FILE__`  for the preset `gap`. It is not tex code.
- Allow to define output files to detect successful runs and do not require empty pdf files for these circumstances.

**Comment by kiryph** [comment]
Regarding question:

> - What did I do incorrectly so that the compilation result is inserted twice?

Adding `do not include pdf` to all environments `\begin{CacheMeCode}` and commands `\cacheMe` removes the duplication.

However, how can I set this in the preset?

**Reply by tobiasBora** [reply]
So first, any option that is added inside `\begin{CacheMe}` can be added in the preset (but order matters, that why it might have failed in your tests, read more for details). So the reason of this double inclusion is that by default we already include `__ROBEXT_OUTPUT_PREFIX__-out.tex` (see my other answer below), and when you add `do not include pdf`, this simply sets `custom include command={}`, effectively removing your own configuration (see that here it would not work if you place this before you own `custom include command`). That being said, you should rather use another name for your file as `-out` is not really made for this purpose (see my other answer for details).

**Comment by tobiasBora** [comment, marked as answer]
Good to see some new applications for this lib! So your issue is that you are using the name `__ROBEXT_OUTPUT_PREFIX__-out.tex`, that is automatically `\input` by default. This file is rather supposed to contain stuff to define new variables rather than things to actually print on the screen (e.g. I use it to recover the depth of the pictures), but not really to typeset stuff, for instance I automatically remove any space added at the end of the file. What I would advice you to do is simply to choose another name, like `__ROBEXT_OUTPUT_PREFIX__-gap.tex`.

**Reply by kiryph** [reply]
Thanks for your helpful comment and answer. Yes, indeed choosing a different name for the generated tex output files, resolved my issue with the duplication.

Now it looks much better.

What I particular like about `robust-externalize` is the possibility to even improve the first run by running the compilation commands in parallel.

Consider this example with 10x `\cacheMe[gap]{code}`.

```tex
\documentclass{article}
\usepackage{robust-externalize}

\robExtConfigure{enable manual mode} % for parallel external runs

\begin{PlaceholderFromCode}{__MY_GAP_TEMPLATE__}
LoadPackage("typeset");;
Typeset(__ROBEXT_MAIN_CONTENT__);
\end{PlaceholderFromCode} 

\robExtConfigure{
  new preset={gap}{
    set compilation command={/usr/local/bin/gap -A -b --nointeract "__ROBEXT_SOURCE_FILE__" > "__ROBEXT_OUTPUT_PREFIX__-gap.tex" && touch "__ROBEXT_OUTPUT_PDF__"},
    set template={__MY_GAP_TEMPLATE__},
    custom include command={%
      \evalPlaceholder{%
        \input{__ROBEXT_CACHE_FOLDER____ROBEXT_OUTPUT_PREFIX__-gap.tex}%
      }%
    },
    print verbatim if no externalization,
  }
} 

\begin{document}

\begin{equation}
  M_1 = \cacheMe[gap]{[ [ 0, 0, 0, 1 ], [ 0, 0, 1, 0 ], [ 0, 1, 0, 0 ], [ 1, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_2 = \cacheMe[gap]{[ [ 0, 0, 0, 2 ], [ 0, 0, 2, 0 ], [ 0, 2, 0, 0 ], [ 2, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_3 = \cacheMe[gap]{[ [ 0, 0, 0, 3 ], [ 0, 0, 3, 0 ], [ 0, 3, 0, 0 ], [ 3, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_4 = \cacheMe[gap]{[ [ 0, 0, 0, 4 ], [ 0, 0, 4, 0 ], [ 0, 4, 0, 0 ], [ 4, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_5 = \cacheMe[gap]{[ [ 0, 0, 0, 5 ], [ 0, 0, 5, 0 ], [ 0, 5, 0, 0 ], [ 5, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_6 = \cacheMe[gap]{[ [ 0, 0, 0, 6 ], [ 0, 0, 6, 0 ], [ 0, 6, 0, 0 ], [ 6, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_7 = \cacheMe[gap]{[ [ 0, 0, 0, 7 ], [ 0, 0, 7, 0 ], [ 0, 7, 0, 0 ], [ 7, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_8 = \cacheMe[gap]{[ [ 0, 0, 0, 8 ], [ 0, 0, 8, 0 ], [ 0, 8, 0, 0 ], [ 8, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_9 = \cacheMe[gap]{[ [ 0, 0, 0, 9 ], [ 0, 0, 9, 0 ], [ 0, 9, 0, 0 ], [ 9, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
  M_{10} = \cacheMe[gap]{[ [ 0, 0, 0, 10 ], [ 0, 0, 10, 0 ], [ 0, 10, 0, 0 ], [ 10, 0, 0, 0 ] ]}
\end{equation}

\end{document}
```

Starting GAP is quite slow (in particular without using a workspace file). But since I have a multi-core Intel i7 with 6 cores, I can actually use them to improve the first compilation time as well:

Clean state:
```
❯ tree .
.
├── README.md
├── robExt-remove-old-figures.py
├── robust-ext-gap-robExt-all-figures.txt
├── robust-ext-gap-robExt-compile-missing-figures.sh
├── robust-ext-gap-robExt-tmp-file-you-can-remove.tmp
├── robust-ext-gap.pdf
├── robust-ext-gap.tex
└── robustExternalize

2 directories, 7 files

❯ time sh -c 'pdflatex robust-ext-gap.tex && cat robust-ext-gap-robExt-compile-missing-figures.sh | sort | uniq | parallel && pdflatex robust-ext-gap.tex'
sh -c   65.54s user 2.94s system 761% cpu 8.989 total
```

It took 9 seconds to complete the first run.

When I clean up, comment the line `\robExtConfigure{enable manual mode}` and use `-shell-escape`, I have to wait 28 seconds:

```
pdflatex --shell-escape robust-ext-gap.tex  26.27s user 1.68s system 97% cpu 28.616 total
```

**Reply by tobiasBora** [reply]
Great you like it! Actually for the parallel version I was thinking to create an option to automatically start it in parallel if, for instance, it sees that more than 5 items are compiled.

And feel free to propose any improvement you can think of, if I have some time I’ll try to implement them. And I’m aware of the issues you mentioned, for now it is a bit dirty to always use a .tex (e.g. I use a .tex file containing python code to run the python preset), but it works for everything I tried and allows simpler code on my side… (e.g. it is trivial to check if a file has been successfully compiled) and if for some cases it is really a problem it is still possible to start the command with a `cp __ROBEXT_OUTPUT_PREFIX__.tex __ROBEXT_OUTPUT_PREFIX__.yourExt`. But I agree it would be nicer to provide a cleaner solution.

**Reply by tobiasBora** [reply]
I just made an option `compile in parallel` and `compile in parallel after=5` to do all this process automatically (`compile in parallel after=5` will only start the compilation in parallel if there are more than 5 elements to build), I was thinking you might be interested.

**Comment by kiryph** [comment]
## Addition: GAP Workspace file to speed up the runs of GAP

Use a dedicated GAP workspace file for the LaTeX document under consideration, e.g. as following.

Create a GAP source code file which creates the workspace file:

```
❯ cat setup-gap-workspace.g
LoadPackage("typeset");
SaveWorkspace("gap-workspace");
```
Here only the `typeset` package is loaded (with its dependent packages). But you can also load further GAP packages and also do some preliminary calculations which only have to be done once for the document.

I use latexmk to create the `gap-workspace` file before the first latex run, by creating a latexmkrc file with a system command

```
❯ cat latexmkrc
system("test ! -e gap-workspace && /usr/local/bin/gap -E --nointeract -A -b setup-gap-workspace.g");
```

The option `-A` surpresses all the default GAP packages.

The demo document uses already the new (November 2023) possibility to activate compilation in parallel within the document:

```
❯ cat robust-ext-gap.tex
\documentclass{article}
\usepackage{robust-externalize}

\robExtConfigure{compile in parallel}

\begin{PlaceholderFromCode}{__MY_GAP_TEMPLATE__}
Typeset(__ROBEXT_MAIN_CONTENT__);
\end{PlaceholderFromCode}

\robExtConfigure{
  new preset={gap}{
    set compilation command={/usr/local/bin/gap --nointeract -A -b -L "__ROBEXT_WAY_BACK__gap-workspace" "__ROBEXT_SOURCE_FILE__" > "__ROBEXT_OUTPUT_PREFIX__-gap.tex" && touch "__ROBEXT_OUTPUT_PDF__"},
    set template={__MY_GAP_TEMPLATE__},
    custom include command={%
      \evalPlaceholder{%
        \input{__ROBEXT_CACHE_FOLDER____ROBEXT_OUTPUT_PREFIX__-gap.tex}%
      }%
    },
    print verbatim if no externalization,
  }
}


\begin{document}

\begin{equation}
	M_1 = \cacheMe[gap]{[ [ 0, 0, 0, 1 ], [ 0, 0, 1, 0 ], [ 0, 1, 0, 0 ], [ 1, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_2 = \cacheMe[gap]{[ [ 0, 0, 0, 2 ], [ 0, 0, 2, 0 ], [ 0, 2, 0, 0 ], [ 2, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_3 = \cacheMe[gap]{[ [ 0, 0, 0, 3 ], [ 0, 0, 3, 0 ], [ 0, 3, 0, 0 ], [ 3, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_4 = \cacheMe[gap]{[ [ 0, 0, 0, 4 ], [ 0, 0, 4, 0 ], [ 0, 4, 0, 0 ], [ 4, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_5 = \cacheMe[gap]{[ [ 0, 0, 0, 5 ], [ 0, 0, 5, 0 ], [ 0, 5, 0, 0 ], [ 5, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_6 = \cacheMe[gap]{[ [ 0, 0, 0, 6 ], [ 0, 0, 6, 0 ], [ 0, 6, 0, 0 ], [ 6, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_7 = \cacheMe[gap]{[ [ 0, 0, 0, 7 ], [ 0, 0, 7, 0 ], [ 0, 7, 0, 0 ], [ 7, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_8 = \cacheMe[gap]{[ [ 0, 0, 0, 8 ], [ 0, 0, 8, 0 ], [ 0, 8, 0, 0 ], [ 8, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_9 = \cacheMe[gap]{[ [ 0, 0, 0, 9 ], [ 0, 0, 9, 0 ], [ 0, 9, 0, 0 ], [ 9, 0, 0, 0 ] ]}
\end{equation}

\begin{equation}
	M_{10} = \cacheMe[gap]{[ [ 0, 0, 0, 10 ], [ 0, 0, 10, 0 ], [ 0, 10, 0, 0 ], [ 10, 0, 0, 0 ] ]}
\end{equation}

\end{document}
```

A fresh compilation now needs only three seconds
```
latexmk -pdf --shell-escape robust-ext-gap.tex  2.69s user 0.51s system 92% cpu 3.461 total
```

When I change a single line of GAP code, the compilation time is now under 1 second

```
latexmk -pdf --shell-escape robust-ext-gap.tex  0.77s user 0.07s system 98% cpu 0.851 total
```

So, using a GAP workspace file is really helpful in case of the GAP system (as is recommended in the documentation).

Be aware of a current issue which might be fixed in the future: https://github.com/gap-system/gap/issues/5014
